### PR TITLE
API Apply Framework\ORM Namespace to model

### LINKED ISSUE #5421: Apply namespaces to the rest of the model subsystem
## Description

The model subsystem has already had namespaces introduced to field types, as part of PHP7 compliance.

Applying namespaces to the rest of the model subsystem is a logical next step.

Note that, as long as we don't yet apply namespaces to any of *the subclasses* of DataObject. #5419 needs to be implemented first.

## Exclusions

 * Updating documentation

## Related

 * Legacy reference (for SilverStripe internal team only, sorry): https://silverstripe.atlassian.net/browse/OSS-1855
 * table_name base implementation: https://github.com/silverstripe/silverstripe-framework/issues/5419

## Tasks

 * [x] Rename `model/` to `orm/`
 * [x] Move out DataObject classes to new `model/` dir (`Image`, `Changeset`)
 * [x] Run upgrader tool against `model/` folder, manually fix any issues which break tests
 * [x] Regression test CMS operation
 * [x] Post PR https://github.com/silverstripe/silverstripe-framework/pull/5734
 * [x] Post to dev list about upcoming change after PR raised but before merging https://groups.google.com/forum/#!topic/silverstripe-committers/t27D84dAdMs https://groups.google.com/forum/#!topic/silverstripe-dev/Ig3ODKBEaT4
 * [ ] Merge PR

Closed with #5734

### PR DESCRIPTION
**Pull request by tractorcow**
See https://github.com/silverstripe/silverstripe-framework/issues/4562 for RFC-6 spec.

Closes #5421

This includes a bunch of other fixes:
- Moves all model classes to a new ORM folder to match the namespace.
- Explicitly require standard sql field methods in DBSchemaManager.
- Add a standard rename map to support the [upgrader](https://github.com/silverstripe/silverstripe-upgrader) tool.

Pull requests for other modules incoming.

- CMS: https://github.com/silverstripe/silverstripe-cms/pull/1534
- SiteConfig: https://github.com/silverstripe/silverstripe-siteconfig/pull/27
- Reports: https://github.com/silverstripe-labs/silverstripe-reports/pull/33
- Asset admin: https://github.com/silverstripe/silverstripe-asset-admin/pull/199
- Test Session: https://github.com/silverstripe-labs/silverstripe-testsession/pull/36 (also https://github.com/silverstripe-labs/silverstripe-testsession/pull/37 to maintain 3.x compat)
- Behat extension: https://github.com/silverstripe-labs/silverstripe-behat-extension/pull/117
- SQlite https://github.com/silverstripe-labs/silverstripe-sqlite3/pull/23

Fixes for other DB connectors will be resolved with a future issue. See https://github.com/silverstripe/silverstripe-framework/issues/5738

### CODE REVIEW

File: cli-script.php [hunk 1]
```diff
@@ -1,5 +1,8 @@
 <?php
 
+use SilverStripe\ORM\DB;
+use SilverStripe\ORM\DataModel;
+
```
Comment (mikenz) [inline cli-script.php]: Also need to remove the
`require_once("model/DB.php");` 
from this file

Comment (tractorcow) [inline cli-script.php]: I've renamed it to "ORM/DB.php" as it currently is named.

**Comment by mikenz** [comment]
Functionally there's no obvious issues for me. The upgrader worked great on all the modules I use and the unit tests all pass when I run them. Nice work!

**Comment by mikenz** [comment]
SilverStripe\ORM\Versioning\Versioned::augmentWrite()
$manipulation uses the table name as the index now, not the class name.

```
framework$ git diff -w
diff --git a/ORM/Versioning/Versioned.php b/ORM/Versioning/Versioned.php
index 3b5297e..16e5dff 100644
--- a/ORM/Versioning/Versioned.php
+++ b/ORM/Versioning/Versioned.php
@@ -896,7 +896,7 @@ class Versioned extends DataExtension implements TemplateGlobalProvider {

             // If we're editing Live, then use (table)_Live instead of (table)
             if($this->hasStages() && static::get_stage() === static::LIVE) {
-                               $this->augmentWriteStaged($manipulation, $class, $id);
+                $this->augmentWriteStaged($manipulation, $table, $id);
             }
         }

```

**Comment by tractorcow** [comment]
Oh, you're right @mikenz good spotting.

**Comment by chillu** [comment]
- We need to adjust heaps of markdown docs - is there a follow up task for this? It's going to be at least 50% manual I'd say.
- We also need to adjust all the [lessons](http://www.silverstripe.org/learn/lessons/), which will be loads of fun (re-recording videos?). Where's that work tracked?
- I think given how many devs will *need* to read the 4.0 upgrading guide closely now, we should revise this document. It's grown quite organically over time, and doesn't have the changes with the biggest impact easily findable (references to the "upgrader" are hidden halfway through the doc)
- We list renamed classes in other parts of the 4.0 upgrading guide, why don't we do this for all the namespaces? It should at least point to `upgrade.yml`.
- PHPDoc follows the [same namespace rules](https://phpdoc.org/docs/latest/guides/types.html#class-name) as PHP which means we don't need to fix `@param`, `@return` and such. But: `{@link MyClassName}` can link to any class (not required to be listed in `use` statements), so we need to go through those. Could be deferred, as long as its tracked in a ticket?
- `i18nTextCollector` will collect fully qualified "namespaces" through `DataObject->provideI18nEntities()` now (for `{$this->class}.SINGULARNAME`. Is it too hacky to use `DataObject::$table_name` for it, assuming that core classes will get a short table name?
- Having an FQN in ModelAdmin URLs is a bit unwieldy, do we need to allow creation of an alias there? E.g. `admin/test/SilverStripe-FrameworkTest-Model-Employee` should be `admin/test/Employee`. Low priority of course.
- Needing to use FQN for string arguments and the `Config` system sounds a pain in the butt. Have aliases been discussed? Sorry, I'm a bit out of the loop on this. Just interested in thinking so far.

**Comment by sminnee** [comment]
> i18nTextCollector will collect fully qualified "namespaces" through DataObject->provideI18nEntities() now (for {$this->class}.SINGULARNAME. Is it too hacky to use DataObject::$table_name for it, assuming that core classes will get a short table name

My vote goes with "too hacky". My preference would be a yml rewriter that we can pass of the lang files - can we do some kind of extract-transform-load step with transifex as a one-off?

**Comment by chillu** [comment]
> can we do some kind of extract-transform-load step with transifex as a one-off

Yeah. It's also going to mess with everybody else's translations though (not just core), so ideally it's an optional part of the upgrader. Using `<MyModel>.SINGULARNAME` will be quite common in projects, because the key is generated by default on text collection. Or more explicitly mentioned in the upgrade guide as a manual task.

**Comment by sminnee** [comment]
> Needing to use FQN for string arguments and the Config system sounds a pain in the butt. Have aliases been discussed? Sorry, I'm a bit out of the loop on this. Just interested in thinking so far.

The one thing that was rejected (I was one of the first people to say it was a bad idea) was just using short-names for yaml config entries, because it defeats the purposes of introducing namespacing.

So, if you were going to have aliasing, you'd need something like a 'use' statement, which would be yet-more-complexity for the config system.

My inclination would be to park it and wait for it to become an issue. I don't think we need to pre-emptively solve problems that might not be that big of an issue.

**Comment by tractorcow** [comment]
> Yeah. It's also going to mess with everybody else's translations though (not just core), so ideally it's an optional part of the upgrader. Using <MyModel>.SINGULARNAME will be quite common in projects, because the key is generated by default on text collection. Or more explicitly mentioned in the upgrade guide as a manual task.

We'll need to rebuild a new set of master keys for 4.0 (treating 3.x and 4.x strings as independent) but that shouldn't be a problem. This is best done near the end of the namespacing. I'll create another story.

**Comment by chillu** [comment]
Happy with the task priorities and split Sam has proposed. +1 from me on merge, I've discovered some missing `DBForeignKey` `use` statements which Damian fixed up now. Trialled this by running the upgrader on `frameworktest`, which has a few models and form fields interacting with them - looks great!

**Comment by kinglozzer** [comment]
> Needing to use FQN for string arguments and the Config system sounds a pain in the butt. Have aliases been discussed? Sorry, I'm a bit out of the loop on this. Just interested in thinking so far.

I’ve encountered this recently, in PHP-land my personal preference is this:

```php
use Config;
use Vendor\Namespace\SubNamespace\ClassName;
// ...
Config::inst()->get(ClassName::class, 'foo');
```

Doesn’t save anything for one-off calls, but saves typing the namespace repeatedly for multiple calls :)

**Comment by sminnee** [comment]
Also, `$this->config()->foo` or `ClassName::config()->foo` works for those classes using the `Configurable` trait.

**Pull request merged**